Corona-Warn-App/src/main/java/de/rki/coronawarnapp/ui/eventregistration/organizer/poster/QrCodePosterViewModel.kt

```diff
@@ -0,0 +1,119 @@
+import timber.log.Timber
+import java.io.File
+import java.io.FileOutputStream
+import java.lang.ref.WeakReference
+
+class QrCodePosterViewModel @AssistedInject constructor(
+    @Assisted private val traceLocationId: Long,
+    private val dispatcher: DispatcherProvider,
+    private val qrCodeGenerator: QrCodeGenerator,
+    private val posterTemplateProvider: PosterTemplateProvider,
+    private val traceLocationRepository: TraceLocationRepository,
+    private val fileSharing: FileSharing
+) : CWAViewModel(dispatcher) {
+
+    private val posterLiveData = MutableLiveData<Poster>()
+    val poster: LiveData<Poster> = posterLiveData
+    val sharingIntent = SingleLiveEvent<FileSharing.FileIntentProvider>()
+
+    init {
+        generatePoster()
+    }
+
+    /**
+     * Create a new PDF file and result is delivered by [sharingIntent]
+     * as a sharing [FileSharing.ShareIntentProvider]
+     */
+    @Suppress("BlockingMethodInNonBlockingContext")
+    fun createPDF(view: View) = launch(context = dispatcher.IO) {
+        try {
+            val weakViewRef = WeakReference(view) // Accessing view in background thread
+            val directory = File(view.context.cacheDir, "poster").apply { if (!exists()) mkdirs() }
+            val file = File(directory, "cwa-qr-code.pdf")
+
+            val weakView = weakViewRef.get() ?: return@launch // View is not existing anymore
+            val pageInfo = PdfDocument.PageInfo.Builder(weakView.width, weakView.height, 1).create()
+
+            PdfDocument().apply {
+                startPage(pageInfo).apply {
+                    weakView.draw(canvas)
+                    finishPage(this)
+                }
+
+                FileOutputStream(file).use {
+                    writeTo(it)
+                    close()
+                }
+            }
+
+            sharingIntent.postValue(fileSharing.getFileIntentProvider(file, traceLocation().description))
+        } catch (e: Exception) {
+            Timber.d(e, "Creating pdf failed")
+            e.report(ExceptionCategory.INTERNAL)
+        }
+    }
+
+    private fun generatePoster() = launch(context = dispatcher.IO) {
+        try {
+            val traceLocation = traceLocation()
+            val template = posterTemplateProvider.template()
+            Timber.d("template=$template")
+            val qrCode = qrCodeGenerator.createQrCode(
+                input = traceLocation.locationUrl,
+                length = template.qrCodeLength,
+                margin = 0
+            )
+
+            val textInfo = buildString {
+                append(traceLocation.description)
+                appendLine()
+                append(traceLocation.address)
+            }
+            posterLiveData.postValue(
+                Poster(qrCode, template, textInfo)
+            )
+        } catch (e: Exception) {
+            Timber.d(e, "Generating poster failed")
+            posterLiveData.postValue(Poster())
```

Comment:
It's a bit confusing to post an empty poster in the error case. Maybe use an `error` Livedata property here as we have it in other `ViewModel`s ?

Reply:
What is the issue with that? the Poster is empty and you can see one line below that the error is reported already. this is just to give a hint to the UI to stop the progress

Reply:
The issue is that it is not consistent with most of the other `ViewModels` in our App, in which we use `errorEvent` properties instead of returning empty data for error cases. Its not a blocker for me, so you can leave it as it is if you want.

Reply:
I think we discussed this before, error reposting in the app is `e.report`